Tdl 14795 streams and sync should not be dict based

dbshah1212 opened the pull request:
# Description of change
- Changed tap structure from dict based streams to class base.
- Updated error handling - removed nested try-catch
- Made bookmark key automatic
- Added end_date param - to handle so many requests
- Fixed pagination issue - Which was falling for child streams
- Added tap tester test cases - As mentioned in the description to test new code changes.

# Manual QA steps
- Older tap behavior should not be changed like state files and catalog files must remain the same.
 - Verify Error handling and backoffs are working properly.
 - Pagination issue got resolved so we should get all records for child streams.
 - Check bookmark keys must be automatic.
 - Tap-tester must be passed.
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

luandy64 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+                # Request dates need to be normalized to project timezone or else errors may occur
+                # Errors occur when from_date is > 365 days ago
+                #   and when to_date > today (in project timezone)
+                from_date = "{}".format(start_window.astimezone(tzone))[0:10]
+                to_date = "{}".format(end_window.astimezone(tzone))[0:10]
```
The `datetime` objects have methods to return just the date portion. So we should use them

```python
>>> start_window = datetime.now()
>>> str(start_window.date())
'2021-09-21'
```



```suggestion
                from_date = str(start_window.astimezone(tzone).date())
                to_date = str(end_window.astimezone(tzone).date())
```

dbshah1212 replied:
Updated

luandy64 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+                #   and when to_date > today (in project timezone)
+                from_date = "{}".format(start_window.astimezone(tzone))[0:10]
+                to_date = "{}".format(end_window.astimezone(tzone))[0:10]
+                LOGGER.info(
+                    "START Sync for Stream: {}{}".format(
+                        self.tap_stream_id,
+                        ", Date window from: {} to {}".format(
+                            from_date, to_date)
+                        if self.bookmark_query_field_from
+                        else "",
+                    )
+                )
```
This is silly to force into one log message

```suggestion
                LOGGER.info("START Sync for Stream: %s", self.tap_stream_id)
                if self.bookmark_query_field_from:
                    LOGGER.info("Date window from: %s to %s", from_date, to_date)
```

dbshah1212 replied:
Updated

luandy64 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+            # funnels and cohorts have a parent endpoint with parent_data and parent_id_field
+            if self.parent_path and self.parent_id_field:
+                # API request data
+                LOGGER.info(
+                    "URL for Parent Stream {}: {}/{}".format(
+                        self.tap_stream_id, self.url, self.parent_path
+                    )
+                )
```
```suggestion
                LOGGER.info("URL for Parent Stream %s: %s/%s", 
                            self.tap_stream_id,
                            self.url,
                            self.parent_path)
```

dbshah1212 replied:
Updated

luandy64 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+
+            for parent_record in parent_data:
+                parent_id = parent_record.get(self.parent_id_field)
+                LOGGER.info('START: Stream: {}, parent_id: {}'.format(
+                    self.tap_stream_id, parent_id))
```
```suggestion
                LOGGER.info('START: Stream: %s, parent_id: %s', self.tap_stream_id, parent_id))
```

dbshah1212 replied:
Updated

luandy64 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+                if self.pagination:
+                    params['page_size'] = limit
+
+                # Poped session_id and page number of last parents stream call.
```
```suggestion
                # Popped session_id and page number of last parents stream call.
```

luandy64 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+                        params['page'] = page
+
+                    # querystring: Squash query params into string and replace [parent_id]
+                    querystring = '&'.join(['%s=%s' % (key, value) for (key, value)
+                                            in params.items()]).replace(
+                        '[parent_id]', str(parent_id))
```
I think the spacing here is hard to read. Breaking it up might make it better

```suggestion
                    querystring = '&'.join(['%s=%s' % (key, value) for (key, value) in params.items()])
                    querystring = querystring.replace('[parent_id]', str(parent_id))
```

dbshah1212 replied:
Updated

luandy64 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+                        self.path,
+                        '?{}'.format(querystring) if querystring else '')
+
+                    LOGGER.info('URL for Stream {}: {}'.format(
+                        self.tap_stream_id, full_url))
```
```suggestion
                    LOGGER.info('URL for Stream %s: %s', self.tap_stream_id, full_url)
```

dbshah1212 replied:
Updated

luandy64 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+                    parent_total, date_total, offset, page, session_id, endpoint_total, max_bookmark_value, total_records = self.get_and_transform_records(
+                        querystring, project_timezone, max_bookmark_value, catalog, last_datetime, endpoint_total, limit, total_records, parent_total, record_count, page, offset, parent_record, date_total)
+                   # End stream != 'export'
+                LOGGER.info('FINISHED: Stream: {}, parent_id: {}'.format(
+                    self.tap_stream_id, parent_id))
```
```suggestion
                LOGGER.info('FINISHED: Stream: %s, parent_id: %s', self.tap_stream_id, parent_id)
```

dbshah1212 replied:
Updated

luandy64 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+                   # End stream != 'export'
+                LOGGER.info('FINISHED: Stream: {}, parent_id: {}'.format(
+                    self.tap_stream_id, parent_id))
+                LOGGER.info(
+                    '  Total records for parent: {}'.format(parent_total))
```
```suggestion
                LOGGER.info('  Total records for parent: %s', parent_total)
```

dbshah1212 replied:
Updated

luandy64 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+                    else "",
+                )
+            )
+            LOGGER.info(
+                "  Total records for date window: {}".format(date_total))
```
```suggestion
            LOGGER.info("  Total records for date window: %s", date_total)
```

dbshah1212 replied:
Updated

luandy64 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+
+            # Update the state with the max_bookmark_value for the stream
+            if bookmark_field:
+                self.write_bookmark(
+                    state, self.tap_stream_id, max_bookmark_value)
```
```suggestion
                self.write_bookmark(state, self.tap_stream_id, max_bookmark_value)
```

dbshah1212 replied:
Updated

luandy64 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+        transformed_data = []
+        for record in data:
+            if record and str(record) != '':
+                # transform reocord and append to transformed_data array
```
```suggestion
                # transform record and append to transformed_data array
```

dbshah1212 replied:
Updated

luandy64 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+        for record in data:
+            if record and str(record) != '':
+                # transform reocord and append to transformed_data array
+                transformed_record = transform_record(record, self.tap_stream_id,
+                                                      project_timezone)
```
```suggestion
                transformed_record = transform_record(record, 
                                                      self.tap_stream_id,
                                                      project_timezone)
```

dbshah1212 replied:
Updated

luandy64 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+                # Check for missing keys
+                for key in self.key_properties:
+                    val = transformed_record.get(key)
+                    if val == '' or not val:
```
```suggestion
                    if not val:
```

luandy64 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+                        bookmark_field=next(iter(self.replication_keys), None),
+                        max_bookmark_value=max_bookmark_value,
+                        last_datetime=last_datetime)
+                    LOGGER.info('Stream {}, batch processed {} records'.format(
+                        self.tap_stream_id, record_count))
```
```suggestion
                    LOGGER.info('Stream %s, batch processed %s records', self.tap_stream_id, record_count)
```

dbshah1212 replied:
Updated

luandy64 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+                bookmark_field=next(iter(self.replication_keys), None),
+                max_bookmark_value=max_bookmark_value,
+                last_datetime=last_datetime)
+            LOGGER.info('Stream {}, batch processed {} records'.format(
+                self.tap_stream_id, record_count))
```
```suggestion
            LOGGER.info('Stream %s, batch processed %s records', self.tap_stream_id, record_count)
```

dbshah1212 replied:
Updated

luandy64 commented on tap_mixpanel/sync.py:
```diff
@@ -468,8 +20,10 @@ def update_currently_syncing(state, stream_name):
+    last_stream = Previous currently synced stream, if the load was interrupted
+    """
     last_stream = singer.get_currently_syncing(state)
     LOGGER.info('last/currently syncing stream: {}'.format(last_stream))
```
```suggestion
    LOGGER.info('last/currently syncing stream: %s', last_stream)
```

dbshah1212 replied:
Updated

luandy64 reviewed (changes requested):
There's a lot to change. Most of it is minor.

I have a few questions scattered throughout that would be nice to have answered.

I still need to review the tests and/or get Kyle S to look at them

luandy64 commented on tap_mixpanel/schema.py:
```diff
@@ -130,10 +130,22 @@ def get_schemas(client, properties_flag):
+
+        mdata = metadata.to_map(mdata)
+
+        if stream_metadata.replication_keys:
+                mdata = metadata.write(
+                    mdata,
+                    ('properties', stream_metadata.replication_keys[0]),
+                    'inclusion',
+                    'automatic')
```
Nevermind, I forget `metadata.get_standard_metadata()` does this already
https://github.com/singer-io/singer-python/blob/0c066de21111d8572425083b4a8792d193c80af1/singer/metadata.py#L41-L42

dbshah1212 commented on .circleci/config.yml:
```diff
@@ -25,6 +25,12 @@ jobs:
+          name: 'Unit Tests'
+          command: |
+            source /usr/local/share/virtualenvs/tap-mixpanel/bin/activate
+            pip install pytest
```
Removed

dbshah1212 commented on tap_mixpanel/client.py:
```diff
@@ -59,35 +60,64 @@ class MixpanelInternalServiceError(MixpanelError):
+}
+
 
 def get_exception_for_error_code(error_code):
     return ERROR_CODE_EXCEPTION_MAPPING.get(error_code, MixpanelError)
```
One ERROR_CODE_EXCEPTION_MAPPING left to remove, removed now.

dbshah1212 commented on tap_mixpanel/__init__.py:
```diff
@@ -32,19 +31,21 @@ def do_discover(client, properties_flag):
-        LOGGER.warning("WARNING: Setting start_date to 1 year ago, {}".format(start_date))
-
+        LOGGER.warning(
+            "WARNING: start_date greater than 1 year maxiumum for API.")
```
Updated

dbshah1212 commented on tap_mixpanel/__init__.py:
```diff
@@ -32,19 +31,21 @@ def do_discover(client, properties_flag):
+        LOGGER.warning(
+            "WARNING: start_date greater than 1 year maxiumum for API.")
+        LOGGER.warning(
+            "WARNING: Setting start_date to 1 year ago, {}".format(start_date))
```
Updated

dbshah1212 commented on tap_mixpanel/client.py:
```diff
@@ -59,35 +60,64 @@ class MixpanelInternalServiceError(MixpanelError):
-    LOGGER.error('ERROR {}: {}, REASON: {}'.format(response.status_code,\
-        response.text, response.reason))
+    LOGGER.error('ERROR {}: {}, REASON: {}'.format(response.status_code,
+                                                   response.text, response.reason))
```
Updated

dbshah1212 commented on tap_mixpanel/client.py:
```diff
@@ -136,15 +166,15 @@ def check_access(self):
             self.disable_engage_endpoint = True
-            LOGGER.warning('Mixpanel returned a 402 from the Engage API. Engage stream will be skipped.')
+            LOGGER.warning(
+                'Mixpanel returned a 402 from the Engage API. Engage stream will be skipped.')
```
Updated

dbshah1212 commented on tap_mixpanel/client.py:
```diff
@@ -59,35 +60,64 @@ class MixpanelInternalServiceError(MixpanelError):
+    429: Server429Error,
     500: MixpanelInternalServiceError}
 
+ERROR_CODE_EXCEPTION_MAPPING = {
```
Yes above one is unused hence removed.

dbshah1212 commented on tap_mixpanel/client.py:
```diff
@@ -167,7 +197,7 @@ def perform_request(self,
                                               **kwargs)
 
-            if response.status_code >= 500:
+            if response.status_code > 500:
```
Because for 500 we handled specific exception `MixpanelInternalServiceError`.

dbshah1212 commented on tap_mixpanel/client.py:
```diff
@@ -261,4 +289,4 @@ def request_export(self, method, url=None, path=None, params=None, json=None, **
             reader = jsonlines.Reader(response.iter_lines())
             for record in reader.iter(allow_none=True, skip_empty=True):
-                yield record
+                yield record
```
added

dbshah1212 commented on tap_mixpanel/client.py:
```diff
@@ -107,9 +137,9 @@ def __enter__(self):
     @backoff.on_exception(backoff.expo,
-                          (Server5xxError, Server429Error, ReadTimeoutError, ConnectionError, Timeout),
+                          (Server5xxError, Server429Error,
+                           ReadTimeoutError, ConnectionError, Timeout),
```
Updated

dbshah1212 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+    def transform_datetime(self, this_dttm):
+        with Transformer() as transformer:
+            new_dttm = transformer._transform_datetime(this_dttm)
+        return new_dttm
```
moved to transform.py

dbshah1212 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+            state["bookmarks"] = {}
+        state["bookmarks"][stream] = value
+        LOGGER.info(
+            "Write state for stream: {}, value: {}".format(stream, value))
```
Updated

dbshah1212 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+            self.url,
+            self.path,
+            '?{}'.format(querystring) if querystring else '')
+        if not data or data is None or data == {} or data == []:
```
Updated

dbshah1212 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+
+            # Cohorts endpoint returns results as a list/array (no data_key)
+            # All other endpoints have a data_key
+            if self.data_key is None or self.data_key == '.':
```
Updated

dbshah1212 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+            for record in data[self.data_key]:
+                # transform reocord and append to transformed_data array
+                transformed_record = transform_record(
+                    record, self.tap_stream_id, project_timezone, parent_record)
```
Updated

dbshah1212 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+                # Check for missing keys
+                for key in self.key_properties:
+                    val = transformed_record.get(key)
+                    if val == '' or not val:
```
Updated

dbshah1212 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+                # End data record loop
+
+            if not transformed_data or transformed_data is None or \
+                    transformed_data == []:
```
Updated

dbshah1212 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+
+            if not transformed_data or transformed_data is None or \
+                    transformed_data == []:
+                LOGGER.info('No transformed data for data = {}'.format(data))
```
Updated

dbshah1212 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+                    max_bookmark_value=max_bookmark_value,
+                    last_datetime=last_datetime)
+                LOGGER.info('Stream {}, batch processed {} records'.format(
+                    self.tap_stream_id, record_count))
```
Updated

dbshah1212 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+                    page,
+                    offset,
+                    to_rec,
+                    total_records))
```
Updated

dbshah1212 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+            if delta_days <= attribution_window:
+                delta_days = attribution_window
+                LOGGER.info("Start bookmark less than {} day attribution window.".format(
+                    attribution_window))
```
Updated

dbshah1212 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+            elif delta_days >= 365:
+                delta_days = 365
+                LOGGER.warning(
+                    "WARNING: Start date or bookmark greater than 1 year maxiumum.")
```
Updated

dbshah1212 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+                LOGGER.warning(
+                    "WARNING: Start date or bookmark greater than 1 year maxiumum.")
+                LOGGER.warning(
+                    "WARNING: Setting bookmark start to 1 year ago.")
```
Updated

dbshah1212 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+
+        return start_window, end_window, days_interval
+
+    def sync(self, state: dict, catalog, config: dict, start_date):
```
Updated

dbshah1212 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+        return start_window, end_window, days_interval
+
+    def sync(self, state: dict, catalog, config: dict, start_date):
+        # the sync method depending on different endpoints
```
Updated cooment

dbshah1212 commented on tap_mixpanel/streams.py:
```diff
@@ -1,125 +1,654 @@
+
+        # Get the latest bookmark for the stream and set the last_integer/datetime
+        last_datetime = None
+        max_bookmark_value = None
```
Yes removed

dbshah1212 commented:
> There's a lot to change. Most of it is minor.
> 
> I have a few questions scattered throughout that would be nice to have answered.
> 
> I still need to review the tests and/or get Kyle S to look at them

@luandy64 In this PR we changed the codebase from dict base to class base so we shuffled most of the code and haven't changed the existing one. 

We updated the code as you suggested and answered your questions. Most of the code wasn't written by us but we learned many things from your comments and noted down those learning so those mistakes won't repeat in upcoming PRs. Thank!! 

Note: The testing part has already been reviewed and approved by Kyle S in the different branch.
PR: https://github.com/singer-io/tap-mixpanel/pull/29
